### PR TITLE
Abstractions for MD5.jl

### PR DESCRIPTION
**Pull request by oxinabox**
@staticfloat

I have split-up some methods: (I think just) `pad_remainder` from `digest!`
and I have added a some methods: (I think just) `buffer_pointer`
and made a few tweaks.

By doing this I can just call SHA.jl from MD5.jl for a lot of the work.

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -142,3 +143,7 @@ show(io::IO, ::SHA3_224_CTX) = write(io, "SHA3 224-bit hash state")
 show(io::IO, ::SHA3_256_CTX) = write(io, "SHA3 256-bit hash state")
 show(io::IO, ::SHA3_384_CTX) = write(io, "SHA3 384-bit hash state")
 show(io::IO, ::SHA3_512_CTX) = write(io, "SHA3 512-bit hash state")
+
+
+# use out types to define a method to get a pointer to the state buffer
```
Comment (staticfloat) [inline src/types.jl]: Is this supposed to say `use our types`?

File: src/types.jl [hunk 2]
```diff
@@ -102,7 +103,7 @@ blocklen(::Type{SHA3_512_CTX}) = UInt64(25*8 - 2*digestlen(SHA3_512_CTX))
 
 
 # short_blocklen is the size of a block minus the width of bytecount
-short_blocklen(::Type{T}) where {T<:Union{SHA1_CTX,SHA2_CTX}} = blocklen(T) - 2*sizeof(state_type(T))
+short_blocklen(::Type{T}) where {T<:SHA_CTX} = blocklen(T) - 2*sizeof(state_type(T))
```
Comment (staticfloat) [inline src/types.jl]: To be honest, I'm not sure `short_blocklen()` should be defined for `SHA3_CTX`.  I guess it doesn't hurt however.

File: src/common.jl [hunk 3]
```diff
@@ -32,9 +32,8 @@ function update!(context::T, data::U) where {T<:Union{SHA1_CTX,SHA2_CTX,SHA3_CTX
     end
 end
 
-
-# Clear out any saved data in the buffer, append total bitlength, and return our precious hash!
-function digest!(context::T) where {T<:Union{SHA1_CTX,SHA2_CTX}}
+# Pad the remainder leaving space for the bitcount
+function pad_remainder!(context::T) where T<:SHA_CTX
```
Comment (staticfloat) [inline src/common.jl]: Splitting this out is a very good change.

**Comment by oxinabox** [comment]
@staticfloat  can tag do a release?
So I can release MD5.jl.

**Comment by staticfloat** [comment]
https://github.com/staticfloat/SHA.jl/releases/tag/v0.5.3

**Pull request merged**